code/game/objects/items/books/skill_book.dm

```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
-		return
+		return FALSE
 	if(!ez_read &&!user.skill_check(SKILL_LITERACY, SKILL_BASIC))
 		to_chat(user, SPAN_WARNING(pick("Haha, you know you can't read. Good joke. Put [title] back.","You open up [title], but there aren't any pictures, so you close it again.","You don't know how to read! What good is this [name] to you?!")))
-		return
+		return FALSE
 
-	if(reading) //Close book, get rid of buffs
-		src.unlearn(user)
-		to_chat(user, SPAN_NOTICE("You close the [name]. That's enough learning for now."))
-		reading = FALSE
-		return
+	if(reading)
+		if(reading != weakref(user))
+			to_chat(user, SPAN_WARNING("\The [reading.resolve()] is already using this textbook!"))
+		else
+			to_chat(user, SPAN_WARNING("You are already using this textbook!"))
+		return FALSE
 
 	if(user.too_many_buffs(/datum/skill_buff/skill_book))
 		to_chat(user, SPAN_WARNING("You can't read two books at once!"))
-		return
+		return FALSE
 
 	if(!user.skill_check(skill, skill_req))
 		to_chat(user, SPAN_WARNING("[title] is too advanced for you! Try something easier, perhaps the \"For Idiots\" edition?"))
-		return
+		return FALSE
+
 	if(user.get_skill_value(skill) > skill_req)
 		to_chat(user, SPAN_WARNING("You already know everything [title] has to teach you!"))
-		return
+		return FALSE
+
+	return TRUE
+
+/obj/item/book/skill/attack_self(mob/user)
+
+	if(reading && reading == weakref(user))
+		//Close book, get rid of buffs
+		src.unlearn(user)
+		to_chat(user, SPAN_NOTICE("You close the [name]. That's enough learning for now."))
+		reading = null
+		STOP_PROCESSING(SSprocessing, src)
+		return TRUE
+
+	if(!check_can_read(user))
+		return TRUE
 
 	to_chat(user, SPAN_NOTICE("You open up the [name] and start reading..."))
-	if(user.do_skilled(4 SECONDS, SKILL_LITERACY, src, 0.75))
-		var/list/buff = list()
-		buff[skill] = 1
-		user.buff_skill(buff, buff_type = /datum/skill_buff/skill_book)
-		reading = TRUE
-		to_chat(user, SPAN_NOTICE("You find the information you need! Better keep the page open to reference it."))
-	else
+	if(!user.do_skilled(4 SECONDS, SKILL_LITERACY, src, 0.75))
 		to_chat(user, SPAN_DANGER("Your perusal of the [name] was interrupted!"))
-		return
+		return TRUE
+
+	if(!check_can_read(user))
+		return TRUE
+
+	var/list/buff = list()
+	buff[skill] = 1
+	user.buff_skill(buff, buff_type = /datum/skill_buff/skill_book)
+	reading = weakref(user)
+	to_chat(user, SPAN_NOTICE("You find the information you need! Better keep the page open to reference it."))
+	START_PROCESSING(SSprocessing, src)
 
 // buff removal
 /obj/item/book/skill/proc/unlearn(var/mob/user)
 	var/list/F = user.fetch_buffs_of_type(/datum/skill_buff/skill_book, 0)
 	for(var/datum/skill_buff/skill_book/S in F)
 		S.remove()
 
+/obj/item/book/skill/Process()
+	if(!reading)
+		return PROCESS_KILL
+	check_buff()
+
+/obj/item/book/skill/proc/check_buff()
+	var/mob/R = reading?.resolve()
+	if(!istype(R) || R.incapacitated())
+		remove_buff()
+		return
+	if(loc != R && !R.Adjacent(src))
```

Comment:
I would at least consider `CanPhysicallyInteract` (possibly for both this and the incapacitated check above).